File: funkotlin/src/main/kotlin/funkotlin/Option.kt
package funkotlin

sealed class Option<out T> {

    data class Some<out T>(val value: T) : Option<T>()
    object None : Option<Nothing>()

    inline fun <R> fold(g: () -> R, f: (T) -> R): R = when (this) {
        is Some -> f(value)
        None -> g()
    }

    inline fun <R> flatMap(f: (T) -> Option<R>): Option<R> = fold(
            { None },
            { f(it) }
    )

    inline fun <R> map(f: (T) -> R): Option<R> = flatMap { Some(f(it)) }

    inline fun foreach(f: (T) -> Unit): Unit = fold({}, f)

    inline fun filter(f: (T) -> Boolean): Option<T> = fold(
            { None },
            { if (f(it)) Some(it) else None }
    )

    inline fun filterNot(f: (T) -> Boolean): Option<T> = filter { !f(it) }

    inline fun exists(f: (T) -> Boolean): Boolean = fold(
            { false },
            { f(it) }
    )

}

inline fun <T> Option<T>.getOrElse(f: () -> T): T = fold(
        { f() },
        { it }
)

inline fun <T> Option<T>.orElse(f: () -> Option<T>): Option<T> = fold(
        { f() },
        { this }
)

inline fun <T> Option<T>.contains(f: () -> T): Boolean = exists { it == f() }

File: scala-playground/src/main/scala/functional/Option.scala
package functional

// Define the usual Option API.
//
// * Constructors (on the object)
// some
// none
// * methods
// map
// flatMap
// orElse
// getOrElse
// any other useful functions
sealed trait Option[+A] {

  @inline def fold[B](n: => B, s: A => B): B = this match {
    case None => n
    case Some(v) => s(v)
  }

  @inline def map[B](f: A => B): Option[B] = fold(None, a => Some(f(a)))

  @inline def flatMap[B](f: A => Option[B]): Option[B] = fold(None, f(_))

  @inline def orElse[B >: A](f: => Option[B]): Option[B] = this match {
    case None => f
    case s@Some(_) => s
  }

  @inline def getOrElse[B >: A](f: => B): B = fold(f, a => a)

  @inline def filter(f: A => Boolean): Option[A] = this match {
    case None => None
    case s@Some(v) => if (f(v)) s else None
  }

  @inline def exists(f: A => Boolean): Boolean = fold(false, f(_))

}

case object None extends Option[Nothing]
final case class Some[A](value: A) extends Option[A]
